# Practica 1 - Parte 2: Programación con Scala en JupyterLab

## Ejercicio 1 — Variables, tipos e inferencia

In [1]:
val nombre: String = "Aitor"
val partidas: Int = 68
var puntuacion: Int = 72
var victorias: Double = 80.4
var activo: Boolean = true

nombre: String = "Aitor"
partidas: Int = 68
puntuacion: Int = 72
victorias: Double = 80.4
activo: Boolean = true

In [2]:
val nombre: String = "María"
val partidas: Int = 67
var puntuacion: Int = 70
var victorias: Double = 78.6
var activo: Boolean = true

nombre: String = "María"
partidas: Int = 67
puntuacion: Int = 70
victorias: Double = 78.6
activo: Boolean = true

In [3]:
println(s"Nombre: $nombre, Partidas: $partidas, Puntuación: $puntuacion, Victorias: $victorias%, Activo: $activo")

Nombre: María, Partidas: 67, Puntuación: 70, Victorias: 78.6%, Activo: true


**Tipos inferidos por Scala**
* **String:** Para el texto (`nombre`).
* **Int:** Para los números enteros (`partidas`, `puntuacion`).
* **Double:** Para los números decimales (`victorias`).
* **Boolean:** Para los valores lógicos (`activo`).

**Diferencias entre versiones**
A nivel interno y de ejecución no existe ninguna diferencia, ambas versiones son estáticamente tipadas. La diferencia es visual y sintáctica: al permitir la inferencia, el código queda mucho más limpio, conciso y fácil de leer, evitando redundancia al no tener que escribir explícitamente el tipo de dato que el compilador ya puede deducir por el valor asignado.

**Uso de `val` (inmutable) y `var` (mutable)**
* **Deberían ser `val`:** El `nombre` del jugador, ya que su identidad no va a cambiar a mitad del juego.
* **Podrían ser `var`:** La `puntuacion`, las `victorias` y el estado `activo` deben ser variables mutables, dado que son propiedades dinámicas que se actualizan continuamente según el desarrollo de la partida. `partidas` podría ser `var` si se incrementa en tiempo real, o `val` si representa un informe final estático.

## Ejercicio 2 — val, var y reasignación

In [4]:
val jugador = "Marta"
var puntuacion = 10

puntuacion += 5

puntuacion += 3

print(puntuacion)

18

jugador: String = "Marta"
puntuacion: Int = 18

In [4]:
jugador = "Marcos"

-- [E052] Type Error: cmd5.sc:1:19 ---------------------------------------------
1 |val res5 = jugador = "Marcos"
  |           ^^^^^^^^^^^^^^^^^^
  |           Reassignment to val jugador
  |
  | longer explanation available when compiling with `-explain`
Compilation Failed

* **Por qué `puntuacion` puede modificarse:** Porque se ha declarado utilizando la palabra reservada `var` (variable), lo que la define como una variable **mutable**. Esto permite que su valor sea reasignado o actualizado (como las sumas += 5 y += 3) durante la ejecución del código.

* **Por qué `jugador` no puede reasignarse:** Porque se ha declarado utilizando la palabra reservada `val` (valor), lo que la convierte en **inmutable**. Una vez inicializada con un dato, no puede volver a apuntar a un valor diferente.

* **Mensaje generado por Scala:** Scala detecta la infracción en tiempo de compilación y bloquea la ejecución lanzando un error de tipo. El mensaje exacto generado es **`Reassignment to val jugador`**, acompañado de un aviso de fallo en la compilación (`Compilation Failed`).

## Ejercicio 3 — Tipos numéricos y precisión

In [5]:
val valorD: Double = 3.14159265358979323846264338327
val valorF: Float =  3.14159265358979323846264338327f

print(valorD)
print(valorF)

var vInt: Int = 23
var vBoolean: Boolean = true
var vString: String = "Tajamar"

3.1415926535897933.1415927

valorD: Double = 3.141592653589793
valorF: Float = 3.1415927F
vInt: Int = 23
vBoolean: Boolean = true
vString: String = "Tajamar"

* **Diferencia entre Double y Float:** La diferencia radica en la precisión. Al imprimir los valores, se observa que el tipo `Double` conserva mucha más precisión en los decimales (`3.141592653589793`), mientras que el tipo `Float` trunca y redondea el valor perdiendo precisión mucho antes (`3.1415927`).
* **Tipo para representar el número de estudiantes:** Utilizaría el tipo `Int` (entero), ya que el número de personas es una cantidad discreta, contable y no tiene partes fraccionarias.
* **Tipo para representar si un estudiante ha aprobado:** Utilizaría el tipo `Boolean`, ya que representa un estado lógico binario (`true` para aprobado, `false` para no aprobado).

## Ejercicio 4 — Función para determinar si una mano se pasa de 21

In [6]:
def bust(puntuacion: Int): Boolean = {
  puntuacion > 21
}

println(bust(18)) 
println(bust(21)) 
println(bust(22)) 
println(bust(30))

false
false
true
true


defined function bust

## Ejercicio 5 — Comparación de dos manos

In [7]:
def maxHand(handA: Int, handB: Int): Int = {
  if (handA > handB) {
    handA
  } else {
    handB
  }
}

println(maxHand(17, 19))
println(maxHand(20, 18))
println(maxHand(21, 21))

19
20
21


defined function maxHand

**Qué sucede cuando ambas manos tienen el mismo valor:**
Al evaluar la condición `if (handA > handB)` con dos valores idénticos (por ejemplo, 21 y 21), la expresión resulta falsa. En consecuencia, la función ignora el bloque `if` y ejecuta el bloque `else`, devolviendo el valor de la segunda variable (`handB`).

**Decisión sobre modificar la función:**
No es necesario modificar el comportamiento de la función. Dado que el objetivo es obtener el valor numérico más alto, devolver `handB` cuando ambos valores son exactamente iguales cumple perfectamente con el propósito de la función, ya que el resultado final devuelto sigue siendo el correcto (21).

## Ejercicio 6 — Decidir el ganador de una partida

In [8]:
def ganador(handA: Int, handB: Int): Int = {
  if (bust(handA) && bust(handB)) {
    0
  } else if (bust(handA)) {
    handB
  } else if (bust(handB)) {
    handA
  } else {
    maxHand(handA, handB)
  }
}

println(s"26 y 20 -> Gana: ${ganador(26, 20)}")
println(s"18 y 22 -> Gana: ${ganador(18, 22)}")
println(s"24 y 25 -> Gana: ${ganador(24, 25)}")
println(s"17 y 19 -> Gana: ${ganador(17, 19)}")
println(s"21 y 20 -> Gana: ${ganador(21, 20)}")

26 y 20 -> Gana: 20
18 y 22 -> Gana: 18
24 y 25 -> Gana: 0
17 y 19 -> Gana: 19
21 y 20 -> Gana: 21


defined function ganador

**Condiciones cumplidas en cada prueba:**

* **26 y 20 (Resultado: 20):** Se cumple la segunda condición (`else if (bust(handA))`), ya que la primera mano supera 21 pero la segunda no.
* **18 y 22 (Resultado: 18):** Se cumple la tercera condición (`else if (bust(handB))`), ya que la segunda mano supera 21 pero la primera se mantiene en el límite.
* **24 y 25 (Resultado: 0):** Se cumple la primera condición del bloque (`if (bust(handA) && bust(handB))`), ya que ambas manos exceden los 21 puntos simultáneamente.
* **17 y 19 (Resultado: 19):** Ninguna mano supera 21, por lo que no se cumple ningún `if` anterior. El flujo cae en el bloque por defecto (`else`), devolviendo la mayor gracias a la función `maxHand`.
* **21 y 20 (Resultado: 21):** Ninguna mano supera 21 (21 es el límite válido exacto y no hace que `bust` devuelva `true`). De nuevo, el flujo cae en la condición por defecto (`else`) y se evalúa cuál es mayor.

## Ejercicio 7 — Arrays y mutabilidad

In [9]:
val jugadores = Array("Alex", "Chen", "Marta")

println(s"Array original: ${jugadores.mkString(", ")}")

jugadores(0) = "Sindhu"
println(s"Array modificado: ${jugadores.mkString(", ")}")

jugadores(1) = 500

-- [E007] Type Mismatch Error: cmd10.sc:8:29 -----------------------------------
8 |val res10_4 = jugadores(1) = 500
  |                             ^^^
  |                             Found:    (500 : Int)
  |                             Required: String
  |
  | longer explanation available when compiling with `-explain`
Compilation Failed

**Por qué los elementos pueden cambiar aunque se use val:**
Al declarar la variable con `val`, lo que se hace inmutable es la referencia; es decir, no puedes hacer que la variable `jugadores` apunte a un array completamente nuevo. Sin embargo, la estructura de datos `Array` en Scala es mutable por naturaleza, lo que permite modificar su contenido interno libremente.

**Por qué no se puede introducir el valor 500:**
Scala es un lenguaje de tipado estático y fuerte. Al inicializar el array solo con textos, el compilador infiere que el tipo exacto es `Array[String]`. Intentar introducir un número entero (`500` es de tipo `Int`) rompe la seguridad de tipos, por lo que el compilador lo bloquea devolviendo un error `type mismatch`.

**Diferencia entre reasignar la variable y modificar el array:**
*   **Reasignar la variable:** Significa intentar que el nombre de la variable apunte a otro espacio de memoria con un objeto distinto (ej. `jugadores = Array("Pedro")`). Esto está prohibido por `val`.
*   **Modificar el array:** Significa acceder al mismo espacio de memoria donde vive el objeto original y alterar sus datos internos (ej. `jugadores(0) = "Sindhu"`). Esto está permitido por la mutabilidad del tipo `Array`.

## Ejercicio 8 — Creación e inicialización de Arrays

In [10]:
val puntuaciones = new Array[Int](4)

println(s"Array inicial: ${puntuaciones.mkString(", ")}")

puntuaciones(0) = 17
puntuaciones(1) = 24
puntuaciones(2) = 21
puntuaciones(3) = 19

println(s"Array modificado: ${puntuaciones.mkString(", ")}")
println(s"Longitud del array: ${puntuaciones.length}")

Array inicial: 0, 0, 0, 0
Array modificado: 17, 24, 21, 19
Longitud del array: 4


puntuaciones: Array[Int] = Array(17, 24, 21, 19)

**Valores de un Array recién creado:**
Cuando se crea un `Array[Int]` en Scala especificando solo su tamaño mediante la palabra reservada `new`, el compilador inicializa automáticamente todas sus posiciones con el valor por defecto del tipo de dato correspondiente. Dado que el tipo es `Int` (entero), el array se llena inicialmente de ceros (`0`), tal y como se observa en la primera impresión por pantalla.

## Ejercicio 9 — Recorrer un Array con while

In [11]:
val manos = Array(17, 24, 21, 19, 26)
var i = 0

while (i < manos.length) {
  val valorActual = manos(i)
  println(s"Mano: $valorActual -> ${bust(valorActual)}")
  i += 1
}

Mano: 17 -> false
Mano: 24 -> true
Mano: 21 -> false
Mano: 19 -> false
Mano: 26 -> true


manos: Array[Int] = Array(17, 24, 21, 19, 26)
i: Int = 5

**Por qué el contador debe ser mutable:**
El contador `i` debe declararse con `var` porque su valor necesita cambiar en cada iteración del bucle (`i += 1`) para avanzar al siguiente índice del array. Si fuera inmutable (`val`), el valor quedaría fijo en 0.

**Qué ocurriría si no incrementaras i:**
La condición de evaluación del bucle `i < manos.length` (es decir, `0 < 5`) siempre se evaluaría como verdadera. Esto crearía un **bucle infinito**, el programa imprimiría el primer elemento ininterrumpidamente y la celda nunca terminaría de ejecutarse hasta bloquear el entorno o ser detenida manualmente.

**Elementos del estilo imperativo:**
*   **Variables mutables:** El uso explícito de un estado mutable (`var i = 0`).
*   **Control de flujo explícito:** El uso de la estructura `while` para decidir cuándo repetir un bloque de código.
*   **Gestión manual de la iteración:** Indicar mecánicamente al programa *cómo* avanzar incrementando el estado interno paso a paso (`i += 1`).
*   **Efectos secundarios:** El uso de `println` dentro del bucle para alterar el estado externo (la pantalla) en cada vuelta, en lugar de devolver un nuevo valor transformado.

## Ejercicio 10 — Listas e inmutabilidad

In [12]:
val jugadores = List("Alex", "Chen", "Marta")
println(s"Lista original: $jugadores")

val nuevaLista = "Sindhu" :: jugadores

println(s"Tras la operación -> Lista original: $jugadores")
println(s"Tras la operación -> Nueva lista: $nuevaLista")

println(s"Longitud de la lista original: ${jugadores.length}")
println(s"Lista invertida: ${jugadores.reverse}")

Lista original: List(Alex, Chen, Marta)
Tras la operación -> Lista original: List(Alex, Chen, Marta)
Tras la operación -> Nueva lista: List(Sindhu, Alex, Chen, Marta)
Longitud de la lista original: 3
Lista invertida: List(Marta, Chen, Alex)


jugadores: List[String] = List("Alex", "Chen", "Marta")
nuevaLista: List[String] = List("Sindhu", "Alex", "Chen", "Marta")

* **Inmutabilidad estricta:** A diferencia de los `Array`, la estructura `List` en Scala es completamente inmutable. Sus elementos no pueden reasignarse ni modificarse una vez creada.
* **Operación `::` (cons):** Como la lista no se puede alterar, al utilizar el operador `::` para añadir un elemento, Scala evalúa la expresión y devuelve una lista completamente nueva. La lista original permanece intacta en memoria.

## Ejercicio 11 — Construcción y concatenación de listas

In [13]:
val lista1 = "Ana" :: "Luis" :: "Marta" :: Nil
val lista2 = List("Pedro", "Sofia")

val listaResultante = lista1 ::: lista2

println(s"Primera lista: $lista1")
println(s"Segunda lista: $lista2")
println(s"Lista resultante: $listaResultante")

Primera lista: List(Ana, Luis, Marta)
Segunda lista: List(Pedro, Sofia)
Lista resultante: List(Ana, Luis, Marta, Pedro, Sofia)


lista1: List[String] = List("Ana", "Luis", "Marta")
lista2: List[String] = List("Pedro", "Sofia")
listaResultante: List[String] = List("Ana", "Luis", "Marta", "Pedro", "Sofia")

**Demostración de inmutabilidad:**
Al imprimir `lista1` y `lista2` después de ejecutar la operación `:::`, se verifica en la salida que ambas mantienen exactamente sus elementos originales. El operador de concatenación no altera las listas existentes, sino que genera y devuelve una tercera lista completamente nueva (`listaResultante`).

## Ejercicio 12 — Operadores relacionales y lógicos

In [14]:
val handA = 18
val handB = 21
val handC = 25

val exp1 = handA > handB
val exp2 = handB == 21
val exp3 = handC != 21
val exp4 = handA <= 21 && handB <= 21
val exp5 = handA > 21 || handC > 21
val exp6 = !(handB > 21)

println(s"¿Es handA mayor que handB?: $exp1")
println(s"¿Es handB igual a 21?: $exp2")
println(s"¿Es handC diferente de 21?: $exp3")
println(s"¿Son handA y handB menores o iguales a 21?: $exp4")
println(s"¿Alguna de las manos handA o handC supera 21?: $exp5")
println(s"¿No se ha pasado handB de 21?: $exp6")

¿Es handA mayor que handB?: false
¿Es handB igual a 21?: true
¿Es handC diferente de 21?: true
¿Son handA y handB menores o iguales a 21?: true
¿Alguna de las manos handA o handC supera 21?: true
¿No se ha pasado handB de 21?: true


handA: Int = 18
handB: Int = 21
handC: Int = 25
exp1: Boolean = false
exp2: Boolean = true
exp3: Boolean = true
exp4: Boolean = true
exp5: Boolean = true
exp6: Boolean = true

| Expresión | Resultado | Explicación |
| :--- | :--- | :--- |
| `handA > handB` | `false` | 18 no es mayor que 21. |
| `handB == 21` | `true` | El valor asignado a `handB` es exactamente 21. |
| `handC != 21` | `true` | El valor de `handC` (25) es distinto de 21. |
| `handA <= 21 && handB <= 21` | `true` | Ambas evaluaciones son verdaderas. El operador lógico AND (`&&`) requiere que las dos condiciones se cumplan simultáneamente. |
| `handA > 21 \|\| handC > 21` | `true` | Aunque `handA` no supera 21, `handC` sí lo hace. El operador lógico OR (`\|\|`) devuelve `true` si al menos una de las condiciones es verdadera. |
| `!(handB > 21)` | `true` | La expresión interna `(handB > 21)` es falsa, ya que 21 no es estrictamente mayor que 21. El operador lógico NOT (`!`) invierte ese resultado, convirtiéndolo en verdadero. |

## Ejercicio 13 — foreach y funciones como valores

In [15]:
val manos = Array(17, 24, 21, 26, 18)

// Versión A: Bucle while
println("Versión A:")
var i = 0
while (i < manos.length) {
  val mano = manos(i)
  println(s"Mano: $mano -> ¿Se pasa?: ${bust(mano)}")
  i += 1
}

println("\nVersión B:")
// Versión B: foreach pasando una función anónima
manos.foreach(mano => println(s"Mano: $mano -> ¿Se pasa?: ${bust(mano)}"))

Versión A:
Mano: 17 -> ¿Se pasa?: false
Mano: 24 -> ¿Se pasa?: true
Mano: 21 -> ¿Se pasa?: false
Mano: 26 -> ¿Se pasa?: true
Mano: 18 -> ¿Se pasa?: false

Versión B:
Mano: 17 -> ¿Se pasa?: false
Mano: 24 -> ¿Se pasa?: true
Mano: 21 -> ¿Se pasa?: false
Mano: 26 -> ¿Se pasa?: true
Mano: 18 -> ¿Se pasa?: false


manos: Array[Int] = Array(17, 24, 21, 26, 18)
i: Int = 5

**Conclusión: Comparación de estilos**

* **¿Cuál necesita un contador?** 
La Versión A (bucle `while`) necesita utilizar un contador explícito (`i`) combinado con la propiedad `manos.length` para saber en qué posición del array se encuentra y cuándo debe detenerse. La Versión B (`foreach`) gestiona todo eso internamente de forma automática.
* **¿Cuál necesita una variable var para recorrer la colección?** 
La Versión A (bucle `while`). Al requerir un contador que se actualice manualmente en cada iteración (`i += 1`), es obligatorio declarar `i` como una variable mutable (`var`).
* **¿Cuál se aproxima más al estilo funcional?** 
La Versión B (`foreach`) se alinea mucho mejor con el estilo funcional de Scala. Oculta la mecánica de la iteración, elimina la necesidad de mutar el estado con variables `var` y demuestra el uso de **funciones de orden superior** al recibir otra función (en este caso, una función anónima) como parámetro para procesar cada elemento.

## Ejercicio 14 — Efectos secundarios y estilo de programación

In [16]:
var total = 0

def sumarAlTotal(valor: Int): Unit = {
  total = total + valor
}

sumarAlTotal(5)
println(s"Total tras la primera llamada: $total")

sumarAlTotal(10)
println(s"Total tras la segunda llamada: $total")

sumarAlTotal(3)
println(s"Total tras la tercera llamada: $total")


def sumar(a: Int, b: Int): Int = {
  a + b
}

val resultado1 = sumar(total, 5)
println(s"\nResultado de función pura sumar($total, 5): $resultado1")

Total tras la primera llamada: 5
Total tras la segunda llamada: 15
Total tras la tercera llamada: 18

Resultado de función pura sumar(18, 5): 23


total: Int = 18
defined function sumarAlTotal
defined function sumar
resultado1: Int = 23

**Explicación y Efecto Secundario:**
* **Por qué modifica una variable externa:** En Scala (y otros lenguajes), las funciones pueden acceder a variables definidas en un ámbito superior (como el entorno global o la clase contenedora). Al estar declarada como `var` (mutable), `sumarAlTotal` tiene permiso para sobreescribir el espacio de memoria de la variable `total`, incluso aunque no se haya pasado como parámetro.
* **Efecto secundario:** El efecto secundario es precisamente la alteración del estado global del programa (la variable `total`). Una función con efectos secundarios es impredecible, ya que ejecutar `sumarAlTotal(5)` no siempre da el mismo resultado global, depende de cuántas veces se haya ejecutado previamente.

**Comparación de funciones:**

| Característica | sumarAlTotal | sumar |
| :--- | :--- | :--- |
| **Modifica datos externos** | Sí | No |
| **Devuelve un resultado calculado** | No (devuelve `Unit`) | Sí (devuelve `Int`) |
| **Utiliza efecto secundario** | Sí | No |
| **Estilo predominante** | Imperativo | Funcional |Ç

## Ejercicio 15 — Programa integrado: torneo de Twenty-One

In [17]:
def bust(puntuacion: Int): Boolean = {
  puntuacion > 21
}

def evaluarRonda(numeroRonda: Int, nombres: List[String], puntuaciones: Array[Int]): Unit = {
  println(s"--- Resultados Ronda $numeroRonda ---")
  var mejorPuntuacion = 0
  var i = 0

  while (i < nombres.length && i < puntuaciones.length) {
    val jugador = nombres(i)
    val mano = puntuaciones(i)

    if (bust(mano)) {
      println(s"$jugador: $mano puntos -> ¡Se pasa!")
    } else if (mano == 21) {
      println(s"$jugador: $mano puntos -> ¡Puntuación perfecta (21)!")
      if (mano > mejorPuntuacion) {
        mejorPuntuacion = mano
      }
    } else {
      println(s"$jugador: $mano puntos -> Puntuación válida.")
      if (mano > mejorPuntuacion) {
        mejorPuntuacion = mano
      }
    }
    i += 1
  }

  if (mejorPuntuacion > 0) {
    println(s">> La mejor puntuación válida de la ronda es: $mejorPuntuacion\n")
  } else {
    println(">> Ningún jugador obtuvo una puntuación válida en esta ronda.\n")
  }
}

val jugadores = List("Alex", "Chen", "Marta", "Sindhu")

// Ronda 1
val manos = Array(18, 24, 21, 20)
evaluarRonda(1, jugadores, manos)

// Ronda 2 (Ampliación)
val manosRonda2 = Array(22, 19, 20, 21)
evaluarRonda(2, jugadores, manosRonda2)

--- Resultados Ronda 1 ---
Alex: 18 puntos -> Puntuación válida.
Chen: 24 puntos -> ¡Se pasa!
Marta: 21 puntos -> ¡Puntuación perfecta (21)!
Sindhu: 20 puntos -> Puntuación válida.
>> La mejor puntuación válida de la ronda es: 21

--- Resultados Ronda 2 ---
Alex: 22 puntos -> ¡Se pasa!
Chen: 19 puntos -> Puntuación válida.
Marta: 20 puntos -> Puntuación válida.
Sindhu: 21 puntos -> ¡Puntuación perfecta (21)!
>> La mejor puntuación válida de la ronda es: 21



defined function bust
defined function evaluarRonda
jugadores: List[String] = List("Alex", "Chen", "Marta", "Sindhu")
manos: Array[Int] = Array(18, 24, 21, 20)
manosRonda2: Array[Int] = Array(22, 19, 20, 21)

**Informe final del ejercicio**

* **Partes mutables:** Las variables declaradas con `var` dentro de la función `evaluarRonda`, concretamente el contador `i` (que aumenta para avanzar en las colecciones) y `mejorPuntuacion` (que se actualiza cada vez que se encuentra un valor válido más alto).
* **Partes inmutables:** Las estructuras de datos principales (`val jugadores`, `val manos`, `val manosRonda2`) permanecen intactas en toda la ejecución. También son inmutables las variables locales de extracción `val jugador` y `val mano` que simplemente leen los datos en cada iteración del bucle.
* **Uso de funciones:** Se ha definido la función pura `bust` para aislar la lógica de evaluación del límite. Además, se ha creado la función `evaluarRonda` para encapsular todo el bloque de procesamiento y permitir su reutilización en diferentes rondas (evitando duplicar el código).
* **Uso de estructuras de control:** Se emplea el bucle `while` para la iteración secuencial a través de los índices de la lista y el array, combinado con un operador lógico `&&` para la condición de salida. Dentro, el bloque `if / else if / else` orquesta las diferentes ramificaciones lógicas dependiendo del valor de la mano actual.
* **Estilo funcional:** Se refleja en la función `bust` (pura, devuelve siempre el mismo booleano para el mismo entero sin modificar nada externo), en la elección de estructuras base inmutables (como la `List` de jugadores declarada con `val`), y en pasar colecciones como parámetros inmutables a las funciones.
* **Estilo imperativo:** Se refleja en el diseño de `evaluarRonda`, ya que utiliza bucles manuales guiados por un estado mutable (`var i`), modifica una variable externa al flujo del bucle (`mejorPuntuacion`) y se apoya fuertemente en efectos secundarios, comunicando los resultados directamente a la pantalla mediante múltiples `println` en lugar de devolver los datos procesados.